to_be_ccppized/coords_1d.F90

```diff
@@ -1,152 +1,154 @@
-      coords%d = d
-    
-      allocate(coords%ifc(coords%n,coords%d+1))
-      allocate(coords%mid(coords%n,coords%d))
-      allocate(coords%del(coords%n,coords%d))
-      allocate(coords%dst(coords%n,coords%d-1))
-      allocate(coords%rdel(coords%n,coords%d))
-      allocate(coords%rdst(coords%n,coords%d-1))
-    
-    end function allocate_coords
-    
-    ! Deallocate and reset to initial state.
-    subroutine finalize(self)
-      class(Coords1D), intent(inout) :: self
-    
-      self%n = 0
-      self%d = 0
-    
-      call guarded_deallocate(self%ifc)
-      call guarded_deallocate(self%mid)
-      call guarded_deallocate(self%del)
-      call guarded_deallocate(self%dst)
-      call guarded_deallocate(self%rdel)
-      call guarded_deallocate(self%rdst)
-    
-    contains
-    
-      subroutine guarded_deallocate(array)
-        real(r8), allocatable :: array(:,:)
-    
-        if (allocated(array)) deallocate(array)
-    
-      end subroutine guarded_deallocate
-    
-    end subroutine finalize
-    
+! This module defines the Coords1D type, which is intended to to cache
+! commonly used information derived from a collection of sets of 1-D
+! coordinates.
+
+use shr_kind_mod, only: r8 => shr_kind_r8
+
+implicit none
+private
+save
+
+public :: Coords1D
+
+type :: Coords1D
+   ! Number of sets of coordinates in the object.
+   integer :: n = 0
+   ! Number of coordinates in each set.
+   integer :: d = 0
+
+   ! All fields below will be allocated with first dimension "n".
+   ! The second dimension is d+1 for ifc, d for mid, del, and rdel, and
+   ! d-1 for dst and rdst.
+
+   ! Cell interface coordinates.
+   real(r8), allocatable :: ifc(:,:)
+   ! Coordinates at cell mid-points.
+   real(r8), allocatable :: mid(:,:)
+   ! Width of cells.
+   real(r8), allocatable :: del(:,:)
+   ! Distance between cell midpoints.
+   real(r8), allocatable :: dst(:,:)
+   ! Reciprocals: 1/del and 1/dst.
+   real(r8), allocatable :: rdel(:,:)
+   real(r8), allocatable :: rdst(:,:)
+ contains
+   procedure :: section
+   procedure :: finalize            ! User-callable
+   final     :: finalize_Coords1D   ! Auto-finalize
+end type Coords1D
+
+interface Coords1D
+   module procedure new_Coords1D_from_fields
+   module procedure new_Coords1D_from_int
+end interface
+
+contains
+
+! Constructor to create an object from existing data.
+function new_Coords1D_from_fields(ifc, mid, del, dst, &
+     rdel, rdst) result(coords)
+  real(r8), USE_CONTIGUOUS intent(in) :: ifc(:,:)
+  real(r8), USE_CONTIGUOUS intent(in) :: mid(:,:)
+  real(r8), USE_CONTIGUOUS intent(in) :: del(:,:)
+  real(r8), USE_CONTIGUOUS intent(in) :: dst(:,:)
+  real(r8), USE_CONTIGUOUS intent(in) :: rdel(:,:)
+  real(r8), USE_CONTIGUOUS intent(in) :: rdst(:,:)
+  type(Coords1D) :: coords
+
+  coords = allocate_coords(size(ifc, 1), size(ifc, 2) - 1)
+
+  coords%ifc = ifc
+  coords%mid = mid
+  coords%del = del
+  coords%dst = dst
+  coords%rdel = rdel
+  coords%rdst = rdst
```

Comment:
@johnmauff - Could you please explain the benefit that removing the `(:,:)` is giving (for our future reference)